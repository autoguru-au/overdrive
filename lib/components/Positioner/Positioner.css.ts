import { style } from '@vanilla-extract/css';

export const root = style({
	left: 0,
	top: 0,
	transform: 'translateZ(0)',
	willChange: 'transform',
	zIndex: 1050,
});
