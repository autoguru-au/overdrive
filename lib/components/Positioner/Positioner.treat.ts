import { style } from 'treat';

export const root = style({
	position: 'fixed',
	zIndex: 20,
	top: 0,
	left: 0,
	transform: 'translateZ(0)',
	willChange: 'transform',
});
