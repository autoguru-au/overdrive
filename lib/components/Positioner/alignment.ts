export enum EAlignment {
	TOP = 'top',
	TOP_LEFT = 'top_left',
	TOP_RIGHT = 'top_right',
	BOTTOM = 'bottom',
	BOTTOM_LEFT = 'bottom_left',
	BOTTOM_RIGHT = 'bottom_right',
	LEFT = 'left',
	RIGHT = 'right',
}
