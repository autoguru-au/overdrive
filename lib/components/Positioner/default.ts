export { Positioner as default } from './Positioner';
