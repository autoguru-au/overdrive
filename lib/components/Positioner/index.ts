export { getOptimalPosition } from './getOptimalPosition';
export { EAlignment as EPositionerAlignment } from './alignment';
export { Positioner } from './Positioner';
