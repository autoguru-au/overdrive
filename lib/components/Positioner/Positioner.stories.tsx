import { Meta, StoryObj } from '@storybook/react-vite';
import isChromatic from 'chromatic/isChromatic';
import React, { useRef } from 'react';

import { Box } from '../Box/Box';
import { Button } from '../Button/Button';
import { Text } from '../Text/Text';

import { Positioner } from './Positioner';
import { EAlignment } from './alignment';

export default {
	title: 'Layout/Positioner',
	component: Positioner,
	parameters: { chromatic: {} },
	argTypes: {
		alignment: {
			options: Object.values(EAlignment),
			defaultValue: EAlignment.BOTTOM_LEFT,
			control: {
				type: 'select',
			},
		},
		triggerOffset: {
			control: {
				type: 'range',
				min: 0,
				max: 400,
			},
		},
	},
} satisfies Meta<typeof Positioner>;

type Story = StoryObj<typeof Positioner>;

export const Closed: Story = {
	args: {
		alignment: EAlignment.BOTTOM_LEFT,
		isOpen: false,
		triggerOffset: 12,
	},
	render: (args) => {
		const triggerRef = useRef(null);
		return (
			<div>
				<Button ref={triggerRef} size="small">
					Open me
				</Button>
				<Positioner {...args} triggerRef={triggerRef}>
					<Box
						boxShadow="1"
						backgroundColour="white"
						borderRadius="1"
						borderWidth="1"
						borderColour="gray"
						padding="2"
					>
						<Text is="p">
							Hello im from the consumer:{' '}
							{isChromatic()
								? '999'
								: Math.ceil(Math.random() * 100)}
						</Text>
					</Box>
				</Positioner>
			</div>
		);
	},
};

export const Open: Story = {
	args: {
		alignment: EAlignment.BOTTOM_LEFT,
		isOpen: true,
		triggerOffset: 12,
	},
	render: (args) => {
		const triggerRef = useRef(null);
		return (
			<div>
				<Button ref={triggerRef} size="small">
					Open me
				</Button>
				<Positioner {...args} triggerRef={triggerRef}>
					<Box
						boxShadow="1"
						backgroundColour="white"
						borderRadius="1"
						borderWidth="1"
						borderColour="gray"
						padding="2"
					>
						<Text is="p">
							Hello im from the consumer:{' '}
							{isChromatic()
								? '999'
								: Math.ceil(Math.random() * 100)}
						</Text>
					</Box>
				</Positioner>
			</div>
		);
	},
};

export const IllustrateAScroll: Story = {
	args: {
		alignment: EAlignment.BOTTOM_LEFT,
		isOpen: true,
		triggerOffset: 12,
	},
	render: (args) => {
		const triggerRef = useRef(null);
		return (
			<div
				style={{
					height: '100%',
					width: '100%',
				}}
			>
				<div
					style={{
						height: 'calc(100vh*5)',
						width: 'calc(100vw*5)',
					}}
				>
					<div
						style={{
							paddingTop: 'calc((100vh*5) / 2)',
							paddingLeft: 'calc((100vw*5) / 2)',
						}}
					>
						<Button ref={triggerRef} size="small">
							I&apos;m the trigger
						</Button>
					</div>
					<Positioner {...args} triggerRef={triggerRef}>
						<Box
							boxShadow="1"
							backgroundColour="white"
							borderRadius="1"
							borderWidth="1"
							borderColour="gray"
							padding="2"
						>
							<Text is="p">
								Hello im from the consumer:{' '}
								{Math.ceil(Math.random() * 100)}
							</Text>
						</Box>
					</Positioner>
				</div>
			</div>
		);
	},
	parameters: {
		chromatic: { disableSnapshot: true },
	},
};
