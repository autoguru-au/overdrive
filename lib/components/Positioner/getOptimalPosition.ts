import { EAlignment } from './alignment';

export interface Dimensions {
	width: number;
	height: number;
}

export interface Position {
	top: number;
	left: number;
}

export interface Rect extends Dimensions, Position {
	right: number;
	bottom: number;
}

export interface AlignmentRect {
	alignment: EAlignment;
	rect: Rect;
}

const DEFAULT_OFFSET = 12;

/**
 * @description
 * The method that will give you the final position, (top, left) values in pixels of where the container needs to be
 *     placed, so that it.
 *     - A: Fit's in the viewport
 *     - B: Places itself at your proffered alignment, if not, it'll be flipped.
 */
export const getOptimalPosition = (
	preferredAlignment: EAlignment,
	triggerRect: Readonly<Rect>, // The rect of the trigger
	containerDimensions: Readonly<Dimensions>, // The width, height of the container
	triggerOffset: number = DEFAULT_OFFSET, // The pixels to leave around the trigger
	viewportOffset: number = DEFAULT_OFFSET, // The pixels to leave around the viewport
	viewportDimensions: Readonly<Dimensions> = getViewportDimensions(),
): AlignmentRect => {
	const { rect, alignment } = getFittedPosition(
		preferredAlignment,
		containerDimensions,
		triggerRect,
		triggerOffset,
		viewportOffset,
		viewportDimensions,
	);

	return {
		alignment,
		rect: handleOverflow(rect, viewportOffset, viewportDimensions),
	};
};

/**
 * @description
 * Given a computed rect, finally we need to handle the overflow viewport "locking" of this rect.
 */
const handleOverflow = (
	rect: Rect,
	viewportOffset: number,
	viewportDimensions: Readonly<Dimensions>,
): Rect => {
	// Push rect to the right if overflowing on the left side of the viewport.
	if (rect.left < viewportOffset) {
		rect.right += Math.ceil(Math.abs(rect.left - viewportOffset));
		rect.left = Math.ceil(viewportOffset);
	}

	// Push rect to the left if overflowing on the right side of the viewport.
	if (rect.right > viewportDimensions.width - viewportOffset) {
		const delta = Math.ceil(
			rect.right - (viewportDimensions.width - viewportOffset),
		);
		rect.left -= delta;
		rect.right -= delta;
	}

	// Push rect down if overflowing on the top side of the viewport.
	if (rect.top < viewportOffset) {
		rect.top += Math.ceil(Math.abs(rect.top - viewportOffset));
		rect.bottom = Math.ceil(viewportOffset);
	}

	// Push rect up if overflowing on the bottom side of the viewport.
	if (rect.bottom > viewportDimensions.height - viewportOffset) {
		const delta = Math.ceil(
			rect.bottom - (viewportDimensions.height - viewportOffset),
		);
		rect.top -= delta;
		rect.bottom -= delta;
	}

	return rect;
};

/**
 * @description
 * Give this method a preferred alignment - and we'll give you back an alignment and rect for where we think the best
 *     place to place it. By best we mean, if the consumer prefers it on the LEFT, but there is no more room left on
 *     the LEFT, we'll place it to right (we flip).
 *
 * The graph here illustrates the difference between horizontal and vertical, its more about on what axis we flip when
 *     there's no more room.
 *
 *  ---------------- TOP ------------------ <- Vertical
 *        +---------------------+
 *        |                     |
 *  LEFT  |       trigger       |  RIGHT -- <- Horizontal
 *        |                     |
 *        +---------------------+
 *  -------------- BOTTOM ----------------- <- Vertical
 */
const getFittedPosition = (
	preferredAlignment: EAlignment,
	containerDimensions: Readonly<Dimensions>,
	triggerRect: Readonly<Rect>,
	triggerOffset: number,
	viewportOffset: number,
	viewportDimensions: Readonly<Dimensions>,
): AlignmentRect =>
	(isAlignedHorizontal(preferredAlignment)
		? handleIsHorizontal
		: handleIsVertical)(
		preferredAlignment,
		viewportOffset,
		viewportDimensions,
		getRectForAlignmentFactory(
			triggerRect,
			triggerOffset,
			containerDimensions,
		),
	);

const handleIsHorizontal = (
	preferredAlignment: EAlignment,
	viewportOffset: number,
	viewportDimensions: Readonly<Dimensions>,
	genRectFn: ReturnType<RectForAlignmentFactoryFn>,
): AlignmentRect => {
	const isFitsOnLeft = (rect: Rect): boolean => rect.left > viewportOffset;
	const isFitsOnRight = (rect: Rect): boolean =>
		rect.right < viewportDimensions.width - viewportOffset;

	const leftRect = genRectFn(EAlignment.LEFT);
	const rightRect = genRectFn(EAlignment.RIGHT);

	const fitsOnLeft = isFitsOnLeft(leftRect);
	const fitsOnRight = isFitsOnRight(rightRect);

	if (preferredAlignment === EAlignment.LEFT) {
		if (fitsOnLeft) {
			return {
				alignment: EAlignment.LEFT,
				rect: leftRect,
			};
		}

		if (fitsOnRight) {
			return {
				alignment: EAlignment.RIGHT,
				rect: rightRect,
			};
		}
	}

	if (preferredAlignment === EAlignment.RIGHT) {
		if (fitsOnRight) {
			return {
				alignment: EAlignment.RIGHT,
				rect: rightRect,
			};
		}

		if (fitsOnLeft) {
			return {
				alignment: EAlignment.LEFT,
				rect: leftRect,
			};
		}
	}

	const spaceRight = Math.abs(
		viewportDimensions.width - viewportOffset - rightRect.right,
	);
	const spaceLeft = Math.abs(leftRect.left - viewportOffset);

	if (spaceRight < spaceLeft) {
		return {
			alignment: EAlignment.RIGHT,
			rect: rightRect,
		};
	}

	return {
		alignment: EAlignment.LEFT,
		rect: leftRect,
	};
};

const handleIsVertical = (
	preferredAlignment: EAlignment,
	viewportOffset: number,
	viewportDimensions: Readonly<Dimensions>,
	genRectFn: ReturnType<RectForAlignmentFactoryFn>,
): AlignmentRect => {
	const isFitsOnTop = (rect: Rect): boolean => rect.top > viewportOffset;
	const isFitsOnBottom = (rect: Rect): boolean =>
		rect.bottom < viewportDimensions.height - viewportOffset;

	const isTop = isAlignedOnTop(preferredAlignment);

	const topRect = genRectFn(
		isTop ? preferredAlignment : flipHorizontal(preferredAlignment),
	);
	const bottomRect = genRectFn(
		isTop ? flipHorizontal(preferredAlignment) : preferredAlignment,
	);

	if (isTop) {
		if (isFitsOnTop(topRect)) {
			return {
				alignment: preferredAlignment,
				rect: topRect,
			};
		}

		if (isFitsOnBottom(bottomRect)) {
			return {
				alignment: flipHorizontal(preferredAlignment),
				rect: bottomRect,
			};
		}
	} else {
		if (isFitsOnBottom(bottomRect)) {
			return {
				alignment: preferredAlignment,
				rect: bottomRect,
			};
		}

		if (isFitsOnTop(topRect)) {
			return {
				alignment: flipHorizontal(preferredAlignment),
				rect: topRect,
			};
		}
	}

	const spaceBottom = Math.abs(
		viewportDimensions.height - viewportOffset - bottomRect.bottom,
	);
	const spaceTop = Math.abs(topRect.top - viewportOffset);

	return spaceBottom < spaceTop
		? {
				alignment: isTop
					? flipHorizontal(preferredAlignment)
					: preferredAlignment,
				rect: bottomRect,
		  }
		: {
				alignment: isTop
					? preferredAlignment
					: flipHorizontal(preferredAlignment),
				rect: topRect,
		  };
};

type RectForAlignmentFactoryFn = (
	triggerRect: Readonly<Rect>,
	triggerOffset: number,
	containerDimensions: Readonly<Dimensions>,
) => (alignment: EAlignment) => Rect;
const getRectForAlignmentFactory: RectForAlignmentFactoryFn = (
	triggerRect,
	triggerOffset,
	containerDimensions,
) => alignment =>
	getRectForAlignment(
		alignment,
		triggerRect,
		containerDimensions,
		triggerOffset,
	);

/**
 * @description
 * Given an alignment, we'll return a "base" rect, for that alignment relative to the trigger's rect.
 */
const getRectForAlignment = (
	alignment: EAlignment,
	triggerRect: Readonly<Rect>,
	containerDimensions: Readonly<Dimensions>,
	triggerOffset: number,
): Rect => {
	const alignedTopY =
		triggerRect.top - containerDimensions.height - triggerOffset;
	const alignedBottomY = triggerRect.bottom + triggerOffset;
	const centerX =
		triggerRect.left +
		triggerRect.width / 2 -
		containerDimensions.width / 2;
	const centerY =
		triggerRect.top +
		triggerRect.height / 2 -
		containerDimensions.height / 2;

	switch (alignment) {
		case EAlignment.TOP_LEFT:
			return makeRect(containerDimensions, {
				top: alignedTopY,
				left: triggerRect.left,
			});
		case EAlignment.TOP:
			return makeRect(containerDimensions, {
				top: alignedTopY,
				left: centerX,
			});
		case EAlignment.TOP_RIGHT:
			return makeRect(containerDimensions, {
				top: alignedTopY,
				left: triggerRect.right - containerDimensions.width,
			});
		case EAlignment.RIGHT:
			return makeRect(containerDimensions, {
				top: centerY,
				left: triggerRect.right + triggerOffset,
			});
		case EAlignment.BOTTOM_RIGHT:
			return makeRect(containerDimensions, {
				top: alignedBottomY,
				left: triggerRect.right - containerDimensions.width,
			});
		case EAlignment.BOTTOM:
			return makeRect(containerDimensions, {
				top: alignedBottomY,
				left: centerX,
			});
		default:
		case EAlignment.BOTTOM_LEFT:
			return makeRect(containerDimensions, {
				top: alignedBottomY,
				left: triggerRect.left,
			});
		case EAlignment.LEFT:
			return makeRect(containerDimensions, {
				top: centerY,
				left:
					triggerRect.left -
					containerDimensions.width -
					triggerOffset,
			});
	}
};

/**
 * @description
 * A helper function, that will "create" a Rect compatible object we use to pass around.
 */
export const makeRect = (
	dimensions: Dimensions,
	position: Position,
): Readonly<Rect> => ({
	...dimensions,
	top: Math.ceil(position.top),
	left: Math.ceil(position.left),
	right: Math.ceil(position.left) + dimensions.width,
	bottom: Math.ceil(position.top) + dimensions.height,
});

const getViewportDimensions = (): Dimensions => ({
	height: document.documentElement.clientHeight,
	width: document.documentElement.clientWidth,
});

const flipHorizontal = (alignment: EAlignment): EAlignment => {
	switch (alignment) {
		case EAlignment.TOP_LEFT:
			return EAlignment.BOTTOM_LEFT;
		case EAlignment.TOP:
		default:
			return EAlignment.BOTTOM;
		case EAlignment.TOP_RIGHT:
			return EAlignment.BOTTOM_RIGHT;
		case EAlignment.BOTTOM_LEFT:
			return EAlignment.TOP_LEFT;
		case EAlignment.BOTTOM:
			return EAlignment.TOP;
		case EAlignment.BOTTOM_RIGHT:
			return EAlignment.TOP_RIGHT;
	}
};

const isAlignedOnTop = (alignment: EAlignment): boolean => {
	switch (alignment) {
		case EAlignment.TOP_LEFT:
		case EAlignment.TOP:
		case EAlignment.TOP_RIGHT:
			return true;
		default:
			return false;
	}
};

const isAlignedHorizontal = (alignment: EAlignment): boolean => {
	switch (alignment) {
		case EAlignment.LEFT:
		case EAlignment.RIGHT:
			return true;
		default:
			return false;
	}
};

// Thanks to the wonderful guys over at Segment, and Evergreen for the technical inspiration for this file. <3
