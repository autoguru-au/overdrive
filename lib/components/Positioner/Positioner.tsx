import * as React from 'react';
import {
	ComponentType,
	FunctionComponent,
	memo,
	ReactChild,
	ReactElement,
	RefObject,
	useLayoutEffect,
	useMemo,
	useRef,
	useState,
} from 'react';
import { useStyles } from 'react-treat';

import { isBrowser } from '../../utils';
import { Modal } from '../Modal';
import { Portal } from '../Portal';
import { EAlignment } from './alignment';
import { AlignmentRect, getOptimalPosition, Rect } from './getOptimalPosition';
import * as styleRefs from './Positioner.treat';

export interface Props {
	alignment?: EAlignment;
	isOpen?: boolean;
	triggerRef: RefObject<HTMLElement>;
	triggerOffset?: number;
	withBackdrop?: boolean;

	onRequestClose?(): void;
}

type WrappedComponent<ExtraProps> = ExtraProps & { triggerRect?: Rect } & Pick<
		Props,
		'isOpen' | 'alignment'
	>;

export function usingPositioner<T extends {} = {}>(
	WrappingComponent: ComponentType<WrappedComponent<T>>,
): FunctionComponent<Props & T> {
	const ReturningComponent: FunctionComponent<Props & T> = ({
		alignment = EAlignment.BOTTOM_LEFT,
		withBackdrop = true,
		isOpen = false,
		onRequestClose = () => void 0,
		triggerRef,
		triggerOffset, // This is defaulted in the getOptimalPosition
		...rest
	}) => {
		if (!isBrowser) return null;

		/* eslint-disable react-hooks/rules-of-hooks */
		const positionerRef = useRef<HTMLDivElement>(null);
		const { alignment: derivedAlignment, rect, triggerRect } =
			usePositionerEffect(
				alignment,
				triggerRef,
				triggerOffset!,
				positionerRef,
				isOpen,
			) ?? {};

		const child = useMemo(() => {
			return isOpen ? (
				<WrappingComponent
					{...(rest as T)}
					alignment={derivedAlignment}
					isOpen={isOpen}
					triggerRect={triggerRect}
				/>
			) : null;
		}, [isOpen, rest]);
		/* eslint-enable react-hooks/rules-of-hooks */

		return withBackdrop ? (
			<Modal
				hideBackdrop
				isOpen={isOpen}
				transition={false}
				onRequestClose={onRequestClose}>
				<PositionerBody
					positionerRef={positionerRef}
					rect={rect}
					child={child}
				/>
			</Modal>
		) : (
			<Portal>
				<PositionerBody
					positionerRef={positionerRef}
					rect={rect}
					child={child}
				/>
			</Portal>
		);
	};

	ReturningComponent.displayName = `usingPositioner(${WrappingComponent.displayName})`;

	return ReturningComponent;
}

const PositionerBody = memo<{
	positionerRef: RefObject<HTMLDivElement>;
	rect?: Rect;
	child: ReactChild | null;
}>(({ rect, positionerRef, child }) => {
	const styles = useStyles(styleRefs);

	return (
		<div
			ref={positionerRef}
			role="none presentation"
			style={{
				visibility:
					positionerRef?.current === null && rect?.left! > 0
						? 'hidden'
						: 'visible',
				...(rect && {
					transform: `translate3d(${rect.left}px, ${rect.top}px, 0px)`,
				}),
			}}
			className={styles.root}>
			{child}
		</div>
	);
});

export const Positioner = usingPositioner(
	({ children }) => children as ReactElement,
);

interface PositionerEffectState extends AlignmentRect {
	triggerRect: Rect;
}

const usePositionerEffect = (
	alignment: EAlignment,
	triggerRef: RefObject<HTMLElement>,
	triggerOffset: number,
	positionerRef: RefObject<HTMLElement>,
	isOpen: boolean,
): PositionerEffectState | null => {
	const [
		positionerResult,
		setPositionerResult,
	] = useState<PositionerEffectState | null>(null);

	useLayoutEffect(() => {
		let current = true;

		if (!triggerRef.current && !positionerRef.current) {
			return void 0;
		}

		let lastFrame = requestAnimationFrame(() => {
			lastFrame = requestAnimationFrame(() => {
				handler();
			});
		});

		function handler() {
			if (
				!current ||
				triggerRef?.current === null ||
				positionerRef?.current === null
			) {
				return;
			}

			if (!isOpen && Boolean(lastFrame)) {
				return cancelAnimationFrame(lastFrame);
			}

			const triggerRect = triggerRef.current!.getBoundingClientRect();
			const containerRect = positionerRef.current!.getBoundingClientRect();

			const height = Math.round(containerRect.height);
			const width = Math.round(containerRect.width);

			setPositionerResult({
				triggerRect,
				...getOptimalPosition(
					alignment,
					triggerRect,
					{
						height,
						width,
					},
					triggerOffset,
				),
			});

			lastFrame = requestAnimationFrame(() => {
				handler();
			});
		}

		return () => {
			current = false;
			if (lastFrame) {
				cancelAnimationFrame(lastFrame);
			}
		};
	});

	return positionerResult;
};
