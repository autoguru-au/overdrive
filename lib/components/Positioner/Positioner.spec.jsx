import { render } from '@testing-library/react';
import * as React from 'react';
import { createRef } from 'react';

import { EAlignment } from './alignment';
import { Positioner } from './Positioner';

describe('<Positioner />', () => {
	it('should not throw', () =>
		expect(() =>
			render(<Positioner triggerRef={createRef()} />),
		).not.toThrow());

	it('should match snapshot', () => {
		const { baseElement } = render(
			<Positioner
				isOpen
				triggerRef={createRef()}
				alignment={EAlignment.BOTTOM}>
				Some body
			</Positioner>,
		);

		expect(baseElement).toMatchSnapshot();
	});

	describe('when #isOpen', () => {
		it('should render the body when open', () => {
			const { baseElement } = render(
				<Positioner
					isOpen
					triggerRef={createRef()}
					alignment={EAlignment.BOTTOM}>
					Some body
				</Positioner>,
			);

			expect(baseElement).toHaveTextContent('Some body');
		});

		it('shouldnt have the body when closed', () => {
			const { baseElement } = render(
				<Positioner
					triggerRef={createRef()}
					alignment={EAlignment.BOTTOM}>
					Some body
				</Positioner>,
			);

			expect(baseElement).not.toHaveTextContent('Some body');
		});
	});
});
