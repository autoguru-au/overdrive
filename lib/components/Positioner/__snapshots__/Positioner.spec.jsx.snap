// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Positioner /> should match snapshot 1`] = `
<body
  style="background-color: rgb(255, 255, 255); color: rgb(33, 35, 56);"
>
  <div />
  <div
    class="base root_default"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="backdropRoot_default_base backdrop_root backdropRoot_invisible_base"
    />
    <div
      class="root"
      role="none presentation"
      style="visibility: visible;"
    >
      Some body
    </div>
  </div>
</body>
`;
